### DISCUSSION TITLE
Regarding airflow db upgrade

### DISCUSSION THREAD

**Discussion by Riturajvats** [apache/airflow#25407] [Q&A]
**Shifting From Airflow 2.2.4 version to 2.3.3 **

When we deploying image in swarm environment , the  worker and scheduler container is not able to run and in the logs of container it's throwing error.

So I want to know is there any way to upgrade the database with specific airflow version outside the container scope like through docker-compose.yaml  as  container is not running so we can't perform airflow db upgrade command inside the container.

![Screenshot (149)](https://user-images.githubusercontent.com/54302152/181816541-9e33b8b7-6ec2-4f28-a91b-ad838637802e.png)

**Comment by jedcunningham** [comment]
To get up and running again, just change one of your containers to run `airflow db upgrade` instead temporarily, then change it back after it's finished running once.

In theory you just need a 2.3.3 pointing at your database to do the upgrade, however it is a good idea to make sure you have the same configuration/etc, so having a container in your setup do it is ideal.

**Comment by potiuk** [comment]
You can take an inspiration (note however that it is quick start so it is not production ready in any way) https://airflow.apache.org/docs/apache-airflow/stable/start/docker.html - there is a manually run init service in docker-compose that you can define in your swarm/compose and run manually when needed. This way you do not have to modify your compose/swarn when you want to run upgrade.

**Reply by Riturajvats** [reply]
airflow-init:
    <<: *airflow-common
    command: version
    environment:
      <<: *airflow-common-env
      _AIRFLOW_DB_UPGRADE: 'true'
      _AIRFLOW_WWW_USER_CREATE: 'true'
      _AIRFLOW_WWW_USER_USERNAME: ${_AIRFLOW_WWW_USER_USERNAME:-}
      _AIRFLOW_WWW_USER_PASSWORD: ${_AIRFLOW_WWW_USER_PASSWORD:-}

this is how  init service is defined in docker-compose., but after using this also no airflow service is coming up due to database error.

**Reply by potiuk** [reply]
Maybe you are upgrading wrong database. You need to debug it.

**Reply by potiuk** [reply]
One of the problems people have - if your Airflow does not get the right SQLALCHEMY connection (especially when it's empty) it will upgrade sqlite local file database. Check that your "init" job is using the right connection (analyse the logs to see it)

**Comment by Riturajvats** [comment]
![Screenshot (151)](https://user-images.githubusercontent.com/54302152/182593347-5955c6da-23f3-46fe-953e-196c82a7aae4.png)
In Init logs it showing like this.
In airflow.cfg file I have given sqlalchemy connecting string value like this 
sql_alchemy_conn = postgresql+psycopg2://airflow:airflow@airflow-postgres/airflow

**Comment by potiuk** [comment]
Have you actually read what the message says? 

![image](https://user-images.githubusercontent.com/595491/182593947-794a2f38-57ef-48ad-bd03-0b850ce6fe70.png)


You know it means that you have to make an action and correct those duplicates? Do you ? Is it not clear?

**Comment by potiuk** [comment]
I think you cannot be clearer than this to be honest.

**Reply by Riturajvats** [reply]
Yes it's quite clear but the problem is no airflow service is up so we can't remove connection by going inside the container . [except for postgres and redis]

**Reply by potiuk** [reply]
Your database is broken. you need to fix it. Airflow does not start BECAUSE your database is broken.

**Reply by potiuk** [reply]
You need to do it manually. That's what the message says. Starting Airflow would not help because it does not work when there are duplicate connection ids - simply because it cannot migrate the database.

**Reply by potiuk** [reply]
And please once you fix it - please make a PR to modify this message in the way that it will be very clear. Can we count on it please? 

Apparently our assumption that users like you will understand this message, were wrong, and you are likely the best person in the world who can figure out how this message should look like so that you have no doubts that you have to fix your database first.

**Reply by Riturajvats** [reply]
Sure no worries will do that once this issue got resolved

**Reply by Riturajvats** [reply]
https://airflow.apache.org/docs/apache-airflow/stable/upgrading-from-1-10/upgrade-check.html#upgrade-check

Found a script to check for upgrading from 1.10.15 to 2.0.0 version but from 2.2.4 to 2.3.3 we're unable to perform  upgrade check. Is this script specific to those which are upgrading from 1.10.15 to 2.0.0 ?

**Reply by potiuk** [reply]
Yes. The upgrade check is 1.10-> 2 only. Generally all 2.* versions are backwards-compatible (we follow semver) so there should be no need to run a check (but you are encouraged to remove deprecations as you see them)

**Reply by potiuk** [reply]
But your case is different - you have broken database simply (likely because someone manually modified it maybe even years ago). There is no upgrade check to prevent it and if you modified the database manually, fixing it manually is the only way you can follow).

With great powers (of modiftying the db) comes great responsibilities (fixing them). Maybe a learning from that for you and your team  is not to modify the db in the future manually.

**Reply by Riturajvats** [reply]
From the database end we are using postgres and in the logs it seems fine and also it is able to connect to airflow database
![Screenshot (155)](https://user-images.githubusercontent.com/54302152/182811851-c2978ffd-7a33-4aba-95bc-3026c73107b2.png)
![Screenshot (154)](https://user-images.githubusercontent.com/54302152/182811861-fed7881a-dc45-4548-bbc0-b30d255cb2c5.png)
.

**Reply by potiuk** [reply]
So what?  did you fix the problem you were pointed at ?

**Reply by potiuk** [reply]
Can you answer if you understand that you need to fix your DB manualy ? Do you understand that you need to remove the duplicates manually? Or is it something still unclear to you?

**Reply by Riturajvats** [reply]
I have removed the duplicate rows  from the database and now it showing quite different init log
![Screenshot (157)](https://user-images.githubusercontent.com/54302152/182814165-88dcc0cb-c75e-4b22-afb4-e055711200fc.png)
.

**Reply by potiuk** [reply]
Then you need to see what it says and solve it. Do you really expect me to hand-hold you with every step? This is not a paid consultancy.

**Reply by potiuk** [reply]
Please do your job first. Honestly, If you are paid for your job you should do it, rather than expect others to do your job. You are expecitng far too much here. Maybe it's not the right job for you if instead of looking and problems and trying to solve them you forward messages back and forth to someone who is trying to help in their free time?

I think you should rethink if you are in the right position.
